lib/recursive-methods.rb

```diff
@@ -1,49 +1,134 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    if n == 1 || n == 0 
+        return 1
+    else 
+        return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n)
 def reverse(s)
```

Comment:
The space complexity here is O(n^2) not O(n) because you're creating a new array with each method call.

If `s[0..-1]`  is O(1), then the time complexity is O(n) for this method.